packages/agent/src/lib/runner.ts

```diff
@@ -0,0 +1,719 @@
+          // TODO: Different agents might structure JSON differently
+          // Need to determine the correct key for the actual content
+          if (this.tool === 'claude') {
+            // TODO: Verify the correct key for Claude's JSON response
+            responseContent =
+              parsedResponse.result ||
+              parsedResponse.content ||
+              parsedResponse.message;
+          } else if (this.tool === 'gemini') {
+            // TODO: Verify the correct key for Gemini's JSON response
+            responseContent =
+              parsedResponse.response ||
+              parsedResponse.content ||
+              parsedResponse.text;
+          }
+        } catch (jsonError) {
+          console.log(
+            colors.yellow(
+              '⚠️  Expected JSON format but got invalid JSON, treating as raw text'
+            )
+          );
+          responseContent = rawOutput;
+        }
+      }
+
+      // Extract string outputs from the response
+      const stringOutputs = this.step.outputs.filter(
+        output => output.type === 'string'
+      );
+      if (stringOutputs.length > 0) {
+        await this.extractStringOutputs(
+          responseContent,
+          stringOutputs,
+          outputs
+        );
+      }
+
+      // Extract file outputs by reading created files
+      const fileOutputs = this.step.outputs.filter(
+        output => output.type === 'file'
+      );
+      if (fileOutputs.length > 0) {
+        await this.extractFileOutputs(fileOutputs, outputs);
+      }
+
+      return { success: true };
+    } catch (error) {
+      return {
+        success: false,
+        error: `Failed to parse outputs: ${error instanceof Error ? error.message : String(error)}`,
+      };
+    }
+  }
+
+  /**
+   * Check if the current tool uses JSON output format
+   */
+  private toolUsesJsonFormat(): boolean {
+    switch (this.tool) {
+      case 'claude':
+      case 'gemini':
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Extract string outputs from the agent response
+   */
+  private async extractStringOutputs(
+    responseContent: string,
+    stringOutputs: Array<{ name: string; description: string }>,
+    outputs: Map<string, string>
+  ): Promise<void> {
+    // Try to parse JSON from the response content if it looks like JSON
+    let jsonData: any = null;
+
+    // Look for JSON block in the response
+    const jsonMatch = responseContent.match(/```json\s*([\s\S]*?)\s*```/);
+    if (jsonMatch) {
+      try {
+        jsonData = JSON.parse(jsonMatch[1]);
+      } catch (error) {
+        console.log(
+          colors.yellow('⚠️  Found JSON block but failed to parse it')
+        );
+      }
+    }
+
+    // If no JSON block found, try parsing the entire response as JSON
+    if (!jsonData) {
+      try {
+        jsonData = JSON.parse(responseContent);
+      } catch (error) {
+        // Not JSON, will need to extract manually
+      }
+    }
```

Comment:
I think it would be better to try this the first time (if it's a valid JSON overall, we just parse that), otherwise we try to match a `json` Markdown block.

This rules out the possibility of misparsing an embedded ```json block within the outer JSON.

Am I missing something?

Reply:
I think this is reasonable. Tools are getting better on supporting JSON, so we can check this case first.